bluesky/callbacks/core.py

```diff
@@ -351,3 +352,46 @@ def start(self, doc):
     def _print(self, out_str):
         self._rows.append(out_str)
         self._out(out_str)
+
+
+class Retrieve(CallbackBase):
+    def __init__(self, root_map=None, handler_reg=None):
+        if root_map is None:
+            root_map = {}
+        if handler_reg is None:
+            handler_reg = {}
+        self.resources = None
+        self.handlers = None
+        self.datums = None
+        self.root_map = root_map
+        self.handler_reg = handler_reg
+
+    def start(self, doc):
+        self.resources = {}
+        self.handlers = {}
+        self.datums = {}
+
+    def resource(self, resource):
+        self.resources[resource['uid']] = resource
+        handler = self.handler_reg[resource['spec']]
+
+        key = (str(resource['uid']), handler.__name__)
```

Comment:
I think I see what you're getting at here, but I'm not sure this is the best way to achieve that. What else could we key on?

Reply:
I pulled this out of the databroker code.

Reply:
https://github.com/NSLS-II/databroker/blob/master/databroker/assets/base_registry.py#L249

Reply:
OK, then let's improve it there as well. I'm concerned that this isn't unique enough. What do you think about:

```py
(str(resource['uid']), handler.__module__, handler.__qualname__)
```

The `__module__` is the one I'm worried about omitting, but we might as well upgrade to `__qualname__` while we're here. Context: https://www.python.org/dev/peps/pep-3155/

Reply:
I like `__qualname__`